Track length

paolafer opened the pull request:
I've realized that the length function in paolina_functions returned the number of voxels instead of the actual length of the track. I have modified that function and added a test to make sure the returned length is the correct one. I have first used the test on the old code, which failed, and then on the new one, which passes.

paolafer committed:
Add test for length function

paolafer committed:
Add missing import

paolafer committed:
Make `length` return geometric rather than topological distance

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -100,8 +100,19 @@ def find_extrema(distance : Dict[Voxel, Dict[Voxel, float]]) -> Tuple[Voxel, Vox
+        raise NoVoxels
+    if len(distance) == 1:
+        return 0.
+    max_distance = 0.
```
This is essentially copy-pasted from `find_extrema`.

We do not do copy-n-paste code reuse in IC: it is an evil practice which leads to confusion and to wasting lots of time and effort.

If you need to reuse some code, wrap it into a convenient abstraction, such as a function, and use that in multiple places.

In this case you should extract the loop from `find_extrema` into a function, and call that function in both `find_extrema` and `length`.

paolafer replied:
Ok.

paolafer replied:
I am also a bit uncomfortable with the argument of length and find_extrema. From an external user's point of view, I would expect to look for the length or the extremes of a track, not of a dictionary of distances. I understand that calling all_pairs_dijkstra_path_length() every time I ask for a length is maybe unnecessarily time-consuming? Is this the reason why the find_extrema function has the overmentioned dictionary as an argument?

jacg replied:
The primary reason is probably decomposition and testability. Optimization might well be a secondary reason.

In `paolina_functions.py` at least, `find_extrema` is *always* called just after `shortest_paths`.

I think it would be reasonable to add a function which calculates the extrema directly from a track: it would simply call `shortest_paths` and `find_extrema` in succession.

You might want to rename `find_extrema` to something more clunky and use the original name for the new one which accepts tracks. And then it would be reasonable to do something similar for `length`.

jacg commented on invisible_cities/reco/paolina_functions_test.py:
```diff
@@ -209,3 +209,27 @@ def test_voxelize_single_hit():
+                  (10,10,11,     1),
+                  (10,10,12,     2),
+                  (10,10,13,     4),
+                  (10,10,14,     8),
```
The 'interesting' energies in `voxel_spec` give the impression that the energies somehow matter in this test. AFAICT, the energies are irrelevant, so don't make them look interesting. It would be best if you could get away with `None` or `'dummy'`. But I suspect that you will need the energies to be

+ numeric
+ greater than zero

paolafer replied:
Yes, they must be positive numbers. What do you suggest?

jacg replied:
1?

paolafer replied:
Ah, ok, I was thinking more about random numbers, but maybe it's an overkill...

jacg replied:
The energy values are completely irrelevant for the test, beyond being nonzero. So I would follow the principle of making them as dummy as possible. But nonzero random numbers would be fine too.

jacg commented on invisible_cities/reco/paolina_functions_test.py:
```diff
@@ -209,3 +209,27 @@ def test_voxelize_single_hit():
+
+    expected_length = 8 + np.sqrt(2)
+
+    assert_almost_equal(track_length, expected_length)
```
As a general point for @everybody-in-IC, I recommend using [`pytest.approx`](https://docs.pytest.org/en/latest/builtin.html#comparing-floating-point-numbers) for doing float comparisons. It now supports numpy arrays (and other containers)!

paolafer committed:
Use approx function when comparing almost equal quantities

paolafer committed:
Apply length and find_extrema functions directly to tracks

paolafer committed:
Set irrelevant energies to 1

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -97,35 +97,31 @@ def shortest_paths(track_graph : Graph) -> Dict[Voxel, Dict[Voxel, float]]:
+    """Find the extrema of a track."""
+    distances = shortest_paths(track)
+    extrema_and_length = find_extrema_and_length(distances)
+    return (extrema_and_length[0], extrema_and_length[1])
```
The last two lines look like a very convoluted way of saying `return find_extrema_and_length(distances)`.

jacg replied:
Actually, no, what I said isn't correct. It's more like `return extrema_and_length[:-1]`.

The fact that I misinterpreted it shows that it needs to be rewritten in a clearer way.

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -97,35 +97,31 @@ def shortest_paths(track_graph : Graph) -> Dict[Voxel, Dict[Voxel, float]]:
-    return  max_distance
+    distances = shortest_paths(track)
+    extrema_and_length = find_extrema_and_length(distances)
+    return extrema_and_length[2]
```
The *magic number* `2`is really unhelpful. I would prefer something like

    _, _, length = find_extrema_and_length(distances)
    return length

jacg committed:
Make `find_extrema` docstring more explicit

jacg committed:
Follow PEP8 on function spacing

jacg committed:
Clarify logic

jacg reviewed (commented):
This PR introduces two main changes

1. `length` calculates the geometric rather than the topological distance between the extrema of a track. 

2. `length` and `find_extrema` accept track graphs, rather than the dictionary of distances between.

... hmm, but before approving, I just want to add a test that ensures that we calculate the length along the track, rather than any shortcut that might be available.

jacg committed:
Test that `length` does not take shortcuts

One part of this test fails. This is because contiguity is
broken. (The author never tested it, so it never worked, and has only
been causing confusion ever since. This is about to be fixed in an
upcoming PR.)

jacg reviewed (approved):
OK, I'm happy with this, apart from the general brokenness of `contiguity` but fixing that does not belong to this PR.
